module/zfs/zvol.c

```diff
@@ -1012,6 +1021,97 @@ zvol_last_close(zvol_state_t *zv)
 	    !(zv->zv_flags & ZVOL_RDONLY))
 		txg_wait_synced(dmu_objset_pool(zv->zv_objset), 0);
 	(void) dmu_objset_evict_dbufs(zv->zv_objset);
+}
+
+/*
+ * Suspend the zvol for recv and rollback.
+ */
+int
+zvol_suspend(const char *name)
+{
+	zvol_state_t *zv;
+	int error = 0;
+
+retry:
+	mutex_enter(&zvol_state_lock);
+	zv = zvol_find_by_name(name);
+	if (zv == NULL) {
+		error = ENOENT;
+		goto out;
+	}
+
+	/* we cannot block in rw_enter if already suspended */
+	if (rw_owner(&zv->zv_suspend_lock) != NULL) {
+		mutex_exit(&zvol_state_lock);
+		schedule_timeout_interruptible(1);
+		goto retry;
+	}
+	/* block all I/O, release in zvol_resume. */
+	rw_enter(&zv->zv_suspend_lock, RW_WRITER);
+
+	if (zv->zv_open_count > 0) {
+		zvol_shutdown_zv(zv);
+	}
+out:
+	mutex_exit(&zvol_state_lock);
+	return (SET_ERROR(error));
+}
+
+int
+zvol_resume(const char *name)
+{
+	zvol_state_t *zv;
+	int error = 0;
+
+	mutex_enter(&zvol_state_lock);
+	zv = zvol_find_by_name(name);
+	if (zv == NULL) {
+		error = ENOENT;
+		goto out;
+	}
+
+	ASSERT(RW_WRITE_HELD(&zv->zv_suspend_lock));
+	if (zv->zv_open_count > 0) {
+		VERIFY0(dmu_objset_hold(name, zvol_tag, &zv->zv_objset));
+		VERIFY3P(zv->zv_objset->os_dsl_dataset->ds_owner, ==, zvol_tag);
```

Comment:
We _could_ start using the `zvol_state_t *` here for `zvol_tag` in the same way a `zfs_sb_t *` is used for the filesystems.  The `zvol_state_t *` is already easily available where we'd need it in `zvol_first_open()`.